kube-runtime/src/reflector/store.rs

```diff
@@ -50,22 +53,20 @@ where
         match event {
             watcher::Event::Applied(obj) => {
                 self.store
+                    .write()
                     .insert(ObjectRef::from_obj_with(obj, self.dyntype.clone()), obj.clone());
             }
             watcher::Event::Deleted(obj) => {
                 self.store
+                    .write()
                     .remove(&ObjectRef::from_obj_with(obj, self.dyntype.clone()));
             }
             watcher::Event::Restarted(new_objs) => {
                 let new_objs = new_objs
                     .iter()
-                    .map(|obj| (ObjectRef::from_obj_with(obj, self.dyntype.clone()), obj))
-                    .collect::<HashMap<_, _>>();
-                // We can't do do the whole replacement atomically, but we should at least not delete objects that still exist
-                self.store.retain(|key, _old_value| new_objs.contains_key(key));
-                for (key, obj) in new_objs {
-                    self.store.insert(key, obj.clone());
-                }
+                    .map(|obj| (ObjectRef::from_obj_with(obj, self.dyntype.clone()), obj.clone()))
+                    .collect::<AHashMap<_, _>>();
+                *self.store.write() = new_objs;
```

Comment:
note that the old version would hash every resource 3x more. this should be notably faster when dealing with large sets of resources.